elytron/WFCORE-4172-tls-1.3.adoc

```diff
@@ -0,0 +1,108 @@
+= WFCORE-4172 Add support for TLS 1.3
+:author:            Farah Juma
+:email:             fjuma@redhat.com
+:toc:               left
+:icons:             font
+:keywords:          comma,separated,tags
+:idprefix:
+:idseparator:       -
+:issue-base-url:    https://issues.jboss.org/browse
+
+== Overview
+
+JDK 11 has added support for TLS 1.3. This task is to update Elytron to be able to make use of TLS 1.3.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFCORE-4172[WFCORE-4172]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-1022[EAP7-1022]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:okotek@redhat.com[Ondrej Kotek]
+
+=== Affected Projects or Components
+
+* WildFly, Security
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* It should be possible to configure a ```server-ssl-context``` and ```client-ssl-context``` in the Elytron subsystem
+using the TLS 1.3 protocol and TLS 1.3 https://tools.ietf.org/html/rfc8446#appendix-B.4[cipher suites]. (Note
+that TLS 1.2 cipher suites cannot be used with TLS 1.3 and vice and versa.)
+
+** A new attribute ```cipher-suite-names``` will be added to both ```server-ssl-context``` and ```client-ssl-context```
+   to specify the TLS 1.3 cipher suites that should be enabled. The format will be a colon (:) separated list of
```

Comment:
In ```cipher-suite-filter``` separator is ```:``` as well as ```,``` . Our documentation use mainly ```,```

Reply:
@fjuma, apologies, maybe you already discussed that with Martin or Adam, but doesn't it make sense to use both `:` and `,` separators for suites separation in newly introduced `cipher-suite-names` to conform current form in already existing `cipher-suite-filter` attribute? This makes it a little bit user-friendly, though.

Reply:
Using just `:` appears to be consistent with OpenSSL: https://wiki.openssl.org/index.php/TLS1.3#Ciphersuites

Reply:
@jstourac As David mentioned, allowing only ```:``` is consistent with the format OpenSSL allows for TLS 1.3 cipher suites.

Reply:
@fjuma, @dmlloyd Thank you. Yes, I know it is consistent with OpenSSL, it is also mentioned in the analysis text. I think that OpenSSL always had just `:` as a separator for cipher-suites even before TSL 1.3 ciphers. But we have always allowed both `:` and `,` in `cipher-suite-filter` although only `:` in newly introduced attribute... which is simply weird... although, if this is the desired approach, then okay. Doc update should be sufficient.

Reply:
@jstourac Prior to TLS 1.3, OpenSSL allowed commas in cipher suite names:

> The cipher list consists of one or more cipher strings separated by colons. Commas or spaces are also acceptable separators but colons are normally used.

Our documentation for the pre-TLS 1.3 ```cipher-suite-filter``` which allows OpenSSL-style cipher list strings is as follows:

> Each item is separated from the other items by a colon (":"), though for compatibility, commas (",") or spaces (" ") are allowed delimiters as well.


For the TLS 1.3 ```cipher-suite-names``` attribute, I think we should stick to ```:``` only. This doesn't affect the pre-TLS 1.3 ```cipher-suite-filter``` attribute.

Reply:
@fjuma ah, sorry I missed that bit. Now I see it in the [manpage](https://www.openssl.org/docs/man1.1.1/man1/ciphers.html) too. Thanks for clarification on this :)